<div class="text-cell-section-header">
<h1 style="text-align: center;"><strong>PROGRAMACI&Oacute;N PARA EL AN&Aacute;LISIS DE DATOS</strong></h1>
</div>
<h1 style="text-align: center;"><strong>Tarea 1 - Principios b&aacute;sicos de programaci&oacute;n cient&iacute;fica.</strong></h1>
<h1 style="text-align: center;"><strong>PRESENTADO POR: DARWIN MERCADO</strong></h1>
<h1 style="text-align: center;"><strong>GRUPO 203008069_3</strong></h1>
<h1 style="text-align: center;"><strong>PRESENTADO A FELIPE ALEXANDER PIPICANO GUZMAN</strong></h1>
<h1 style="text-align: center;"><strong>Universidad Nacional Abierta y a Distancia &ndash; UNAD</strong></h1>
<h1 style="text-align: center;"><strong>18 de septiembre del 2023</strong></h1>

# **Presentación del ejercicio**

En este ejercicio se utilizaran varios operadores, matrices y vectores, para calcular las siguientes solicitudes por parte del tutor.

• Calcule el siguiente vector x, usando operaciones de matrices
(dot, transpose, inv) de numpy.linalg:
$ = (𝐴^
𝑇𝐴)
−1𝐴^
𝑇𝑏$
y calcule $‖𝐴𝑥 − 𝑏‖$ utilizando la función numpy.linalg.norm.

• Solucione el siguiente sistema de ecuaciones, mediante la
función numpy.linalg.solve (de ser necesario, investigue la
función en la documentación de numpy):
$(𝐴^
𝑇𝐴)𝑥 = 𝐴^
𝑇𝑏$
y calcule $‖𝐴𝑥 − 𝑏‖$ utilizando la función numpy.linalg.norm.

• Solucione el siguiente sistema de ecuaciones, mediante la
función numpy.linalg.lstsq (de ser necesario, investigue la
función en la documentación de numpy):
$𝐴𝑥 = 𝑏$
y calcule $‖𝐴𝑥 − 𝑏‖$ utilizando la función numpy.linalg.norm.
5

• Compare todos los resultados obtenidos y anexe una discusión
en el análisis de resultados.


# **Revisión teórica**

- **Arreglos**: Un arreglo es una estructura de datos que almacena datos del mismo tipo, ya sean numeros, cadenas de texto, entre otros. para acceder a sus elementos es mediante un indice el cual inicia en 0. (Thakur, 2016)

- **Vectores**: Los vectores es una secuencia ordena de numeros o elementos y tienen una estructura unidimensional, estos son esenciales en algebra lineal para representar puntos, fuerzas, entre otras caracteristicas. (Thakur, 2016)

- **Matriz**: Es una estructura bidimensional que cuenta con filas y columnas, los cuales cada uno tiene su indice, entre sus usos mas importantes se destaca el uso en ecuaciones lineales, estadistica y transformación de algunos datos. (Thakur, 2016)

- **Operadores matemáticos**: Los operadores matemáticos son herramientas claves para la realización de calculos y exploración de ecuaciones, su uso es importante en varios campos, tales como la fisica, quimica, matemática, ciencia de datos, entre otros, los más utilizados son:
Suma (+)
Resta(-)
Multiplicación(*)
División(/)
Producto escalar.  (Thakur, 2016)

# **Desarrollo del código**

In [ ]:
import numpy as np

In [ ]:
#Construya la siguiente matriz como un arreglo de numpy:
A = np.array([[13, 1],
              [14, 1],
              [15, 1],
              [16, 1],
              [17, 1],
              [18, 1]])

print(A)

[[13  1]
 [14  1]
 [15  1]
 [16  1]
 [17  1]
 [18  1]]


In [ ]:
##Construya el siguiente vector como arreglo de numpy
B = np.array([[16],[24],[45],[57],[75],[102]])
B

array([[ 16],
       [ 24],
       [ 45],
       [ 57],
       [ 75],
       [102]])

# $ x = (𝐴^𝑇𝐴)^-1 𝐴^𝑇𝑏$

In [ ]:
ATA = np.dot(A.T, A)
ATB= np.dot(A.T, B)
x=np.dot(np.linalg.inv(ATA),ATB)

In [ ]:
x

array([[  17.        ],
       [-210.33333333]])

# $ ||Ax-b||$

In [ ]:
norm1=np.linalg.norm(np.dot(A,x)-B)

In [ ]:
norm1

10.832051206181278

# $(𝐴^𝑇𝐴)𝑥 = 𝐴^𝑇𝑏$

In [ ]:
ATA = np.dot(A.T, A)
ATB= np.dot(A.T, B)
w = np.linalg.solve(ATA,ATB)
w

array([[  17.        ],
       [-210.33333333]])

# $ ||Ax-b||$

In [ ]:
a = np.linalg.norm(np.dot(A,w)-B)
a

10.832051206181278

# $𝐴𝑥 = 𝑏$

In [ ]:
z= np.linalg.lstsq(A,B, rcond=None)[0]

In [ ]:
z

array([[  17.        ],
       [-210.33333333]])

# $ ||Ax-b||$

In [ ]:
norm=np.linalg.norm(np.dot(A,z)-B)

In [ ]:
norm

10.832051206181276

# Análisis de los resultados

## $ x = (𝐴^𝑇𝐴)^-1 𝐴^𝑇𝑏$

- En el primer calculo de vector de x se utilizo las funciones de la libreria numpy.np.dot, la cual permite realizar el producto escalar entre dos arreglos o vectores. También se utilizo la función .T que es la traspuesta que viene en python como configuración básica, con estas funciones se calculó el valor de $A^T*A$ y el $A^T * b$ de manera separada, para luego utilizar np.linalg.inv que nos permite calcular la inversa, en este caso de el producto $A^T*A$, para así obtener el valor de x. Con esto tenemos una ecuación lineal $ y = 17*x - 210.3333$, que se podría graficar bajo los puntos dados inicialmente y sería nuestra recta ideal a partir de un gráfico de dispersión.

## $ ||Ax-b||$

Luego se obtuvo la norma de $ ||Ax-b||$ a partir del uso de numpy.linalg.norm. con la cual pudimos mirar la medida del tamaño o longitud en el espacio con un valor de 10.832051206181278.

## $(𝐴^𝑇𝐴)𝑥 = 𝐴^𝑇b$

Este sistema de ecuaciones se solucionó a partir del uso de la función numpy.linalg.solve la cual nos permite "Resolver una ecuación matricial lineal o un sistema de ecuaciones escalares lineales" obtenido de Numpy Documentation. Al usar esta función vemos que nos da la misma ecuación que se obtuvo en el primer calculo que se hizo de x que es de $ y = 17*x - 210.3333$ dado que esta resuelve un sistema de ecuaciones y nos da el resultado de $y=mx + b$ que es la formula de una ecuación lineal, como los datos eran los mismos se obtuvo el mismo valor de y.

## $ ||Ax-b||$
Se vuelve a obtener la norma con el uso de numpy.linalg.norm que como ya se había mencionado antes es el tamaño o longitud en el espacio y el valor es el mismo, dado que las variables son las mismas y se usa la misma ecuación.

## $ Ax=b$
Este sistema de ecuaciones se solucionó a partir de numpy.linalg.lstsq la cual nos devuelve la solucion de minimos cuadrados para una ecuación lineal. Se observo que se le debe de indicar que nos muestre la posición cero al finalizar nuestro código para que aí solamente nos muestre la ecuación lineal $ y = 17*x - 210.3333$, que es el mismo resultado como es de esperar dado que los datos inicialmente dados siguen siendo los mismos y se está utilizando una metodología diferente para calcular la misma ecuación lineal.

## $ ||Ax-b||$

Se vuelve a obtener la norma a partir del uso de la misma función numpy.linalg.norm la cual nos va a dar el mismo resultado, dado que la ecuación sigue teniendo los mismos valores.

# **Conclusiones**

Se pudo observar que la ecuación obtenida fue la misma en todas las formas de ecuación, dado que en cada uno de los procedimientos se utilizaba la misma ecuación, simplemente tenían diferentes posiciones en su calculo, lo cual nos llevaba siempre al mismo resultado. si observamos la ecuación de $ x = (𝐴^𝑇𝐴)^-1 𝐴^𝑇𝑏$ podemos observar que se está despejando x, pero si en cambio devolvemos la ecuación a este estado $(𝐴^𝑇𝐴)𝑥 = 𝐴^𝑇b$ simplemente se esta quitando la inversa, con lo cual podemos cancelar las A^T y obtener la última ecuación solicitada de $Ax=b$ con lo cual se obtiene un mismo resultado usando los diferentes metodos en python con la librería de numpy. Asi mismo con la norma esta en las que fueron solicitadas por el tutor siempre tenían la misma forma, por lo tanto se obtenía la norma de 10.832051206181278. en todos los resultados.

# **Referencias**

- Álvarez, C. (2020). Introducción al Jupyter Notebook y aplicaciones básicas. https://repository.unad.edu.co/handle/10596/35681

- Thakur, A. (2016). Python: Real-World Data Science. Packt Publishing. https://bibliotecavirtual.unad.edu.co/login?url=https://search.ebscohost.com/login.aspx?direct=true&db=nlebk&AN=1250754&lang=es&site=ehost-live&ebv=EB&ppid=pp_199

- NumPy. (s.f.). Numpy. Obtenido de https://numpy.org/doc/stable/reference/generated/numpy.linalg.lstsq.html.

- Numpy. (s.f.). Numpy. Obtenido de https://numpy.org/doc/stable/reference/generated/numpy.linalg.solve.html.

- NumPy. (s.f.). NumPy. Obtenido de https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.htm